libkbfs/block_disk_store.go

```diff
@@ -0,0 +1,525 @@
+// Copyright 2016 Keybase Inc. All rights reserved.
+// Use of this source code is governed by a BSD
+// license that can be found in the LICENSE file.
+
+package libkbfs
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/keybase/kbfs/kbfscodec"
+	"github.com/keybase/kbfs/kbfscrypto"
+)
+
+// blockDiskStore stores block data in flat files on disk.
+//
+// The directory layout looks like:
+//
+// dir/0100/0...01/data
+// dir/0100/0...01/id
+// dir/0100/0...01/ksh
+// dir/0100/0...01/refs
+// ...
+// dir/01cc/5...55/id
+// dir/01cc/5...55/refs
+// ...
+// dir/01dd/6...66/data
+// dir/01dd/6...66/id
+// dir/01dd/6...66/ksh
+// ...
+// dir/01ff/f...ff/data
+// dir/01ff/f...ff/id
+// dir/01ff/f...ff/ksh
+// dir/01ff/f...ff/refs
+//
+// Each block has its own subdirectory with its ID truncated to 17
+// bytes (34 characters) as a name. The block subdirectories are
+// splayed over (# of possible hash types) * 256 subdirectories -- one
+// byte for the hash type (currently only one) plus the first byte of
+// the hash data -- using the first four characters of the name to
+// keep the number of directories in dir itself to a manageable
+// number, similar to git.
+//
+// Each block directory has the following files:
+//
+//   - id:   The full block ID in binary format. Always present.
+//   - data: The raw block data that should hash to the block ID.
+//           May be missing.
+//   - ksh:  The raw data for the associated key server half.
```

Comment:
Since I'm here, I changed key_server_half to ksh to shave off a few more path characters.